pkg/roachprod/vm/gce/gcloud.go

```diff
@@ -1178,74 +1223,227 @@ func (p *Provider) Create(
-
-	var labelPairs []string
-	addLabel := func(key, value string) {
-		labelPairs = append(labelPairs, fmt.Sprintf("%s=%s", key, value))
-	}
-
-	for key, value := range opts.CustomLabels {
-		_, ok := m[strings.ToLower(key)]
-		if ok {
-			return fmt.Errorf("duplicate label name defined: %s", key)
-		}
-		addLabel(key, value)
-	}
-	for key, value := range m {
-		addLabel(key, value)
-	}
-	labels := strings.Join(labelPairs, ",")
-
-	args = append(args, "--labels", labels)
 	args = append(args, "--metadata-from-file", fmt.Sprintf("startup-script=%s", filename))
 	args = append(args, "--project", project)
 	args = append(args, fmt.Sprintf("--boot-disk-size=%dGB", opts.OsVolumeSize))
-	var g errgroup.Group
+	return args, cleanUpFn, nil
+}
 
-	nodeZones := vm.ZonePlacement(len(zones), len(names))
-	// N.B. when len(zones) > len(names), we don't need to map unused zones
-	zoneToHostNames := make(map[string][]string, min(len(zones), len(names)))
-	for i, name := range names {
-		zone := zones[nodeZones[i]]
-		zoneToHostNames[zone] = append(zoneToHostNames[zone], name)
+func instanceTemplateName(clusterName string) string {
+	return fmt.Sprintf("%s-template", clusterName)
+}
+
+func instanceGroupName(clusterName string) string {
+	return fmt.Sprintf("%s-group", clusterName)
+}
+
+// createInstanceTemplate creates an instance template for the cluster. This is
+// currently only used for managed instance group clusters.
+func createInstanceTemplate(clusterName string, instanceArgs []string, labelsArg string) error {
+	templateName := instanceTemplateName(clusterName)
+	createTemplateArgs := []string{"compute", "instance-templates", "create"}
+	createTemplateArgs = append(createTemplateArgs, instanceArgs...)
+	createTemplateArgs = append(createTemplateArgs, "--labels", labelsArg)
+	createTemplateArgs = append(createTemplateArgs, templateName)
+
+	cmd := exec.Command("gcloud", createTemplateArgs...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "Command: gcloud %s\nOutput: %s", createTemplateArgs, output)
 	}
-	l.Printf("Creating %d instances, distributed across [%s]", len(names), strings.Join(zones, ", "))
+	return nil
+}
 
-	for zone, zoneHosts := range zoneToHostNames {
-		argsWithZone := append(args[:len(args):len(args)], "--zone", zone)
-		argsWithZone = append(argsWithZone, zoneHosts...)
+// createInstanceGroups creates an instance group for the cluster in each zone
+// specified. Instance groups will be created for the specified zones even if no
+// instances are created in that zone during cluster creation.
+func createInstanceGroups(project, clusterName string, zones []string, opts vm.CreateOpts) error {
+	groupName := instanceGroupName(clusterName)
+	templateName := instanceTemplateName(clusterName)
+	createGroupArgs := []string{"compute", "instance-groups", "managed", "create",
+		"--template", templateName,
+		"--size", "0",
+		"--stateful-external-ip", "enabled",
+		"--stateful-internal-ip", "enabled",
```

Comment:
Why do we need the IPs to be stateful? Might be good to include this in the comments too.